src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.test.ts

```diff
@@ -0,0 +1,282 @@
+  lastModified: new Date().getTime(),
+  cloudProvider
+})
+
+const baseTestFile: AnalysisFile = getTestFile('test/file0.ipynb' as AbsolutePath)
+
+const onDismiss = jest.fn()
+const onSuccess = jest.fn()
+
+const defaultModalProps: AnalysisDuplicatorProps = {
+  destroyOld: false,
+  fromLauncher: false,
+  printName: baseTestFile.fileName,
+  toolLabel: toolLabelTypes[tools.Jupyter.label],
+  workspaceInfo: defaultGoogleWorkspace.workspace,
+  onDismiss, onSuccess
+}
+
+describe('AnalysisDuplicator', () => {
+  beforeEach(() => {
+    // Arrange
+    const googleStorageMock: Partial<GoogleStorageContract> = ({
+      listAnalyses: jest.fn(() => Promise.resolve([]))
+    })
+    asMockedFn(GoogleStorage).mockImplementation(() => googleStorageMock as GoogleStorageContract)
+
+    const azureStorageMock: Partial<AzureStorageContract> = ({
+      listNotebooks: jest.fn()
+    })
+    asMockedFn(AzureStorage).mockImplementation(() => azureStorageMock as AzureStorageContract)
+
+    asMockedFn(useAnalysisFiles).mockImplementation(() => ({
+      analyses: [],
+      refresh: () => {},
+      loading: false
+    }))
+  })
+
+  it('renders correctly by default with destroyOld false', () => {
+    render(h(AnalysisDuplicator, defaultModalProps))
+
+    screen.getByText(`Copy "${baseTestFile.fileName}"`)
+  })
+
+  it('renders correctly by default with destroyOld true', () => {
+    render(h(AnalysisDuplicator, {
+      ...defaultModalProps,
+      destroyOld: true
+    }))
+
+    screen.getByText(`Rename "${baseTestFile.fileName}"`)
+  })
+
+  it.each(
+    [
+      { inputText: ' ', errorMsg: 'Name can\'t be blank' },
+      { inputText: 'invalid$', errorMsg: 'Name can\'t contain these characters:' },
+      { inputText: 'file1', errorMsg: 'Name already exists' }
+    ]
+  )('rejects existing and invalid names', async ({ inputText, errorMsg }) => {
+    const fileList = [getTestFile('test/file1.ipynb' as AbsolutePath), getTestFile('test/file2.ipynb' as AbsolutePath)]
+    asMockedFn(useAnalysisFiles).mockImplementation(() => ({
+      analyses: fileList,
+      refresh: () => {},
+      loading: false
+    }))
+
+    act(() => {
+      render(h(AnalysisDuplicator, defaultModalProps))
+    })
+
+    const input = screen.getByLabelText(/New Name/)
+    await userEvent.type(input, inputText)
+
+    expect(screen.getAllByText(errorMsg).length).toBe(2)
+    const button = screen.getByText('Copy Analysis')
+    expect(button).toHaveAttribute('disabled')
+  })
+
+  it('copies for a google workspace correctly', async () => {
+    const fileList = [getTestFile('test/file1.ipynb' as AbsolutePath), getTestFile('test/file2.ipynb' as AbsolutePath)]
+    const calledMock = jest.fn()
+    const analysisMock = jest.fn(() => ({
+      copy: calledMock
+    }))
+    const googleStorageMock: Partial<GoogleStorageContract> = ({
+      listAnalyses: () => Promise.resolve(fileList),
+      // @ts-expect-error
```

Comment:
What's the error here? As much as possible, we'd like to avoid ignoring type checks in mocks.

Reply:
Its unhappy here because `analysis` is a complex function that takes some arguments and returns a set of properties defined within the `GoogleStorage` object... not super sure the syntactic wrangling I'd need to do to resolve it, the typescript error message is very unhelpful

![image](https://user-images.githubusercontent.com/6465084/202021914-54f33cb0-7f9f-4c58-9c57-aaa35a350b58.png)

Reply:
Wondering if this needs a cast `as Partial<GoogleStorageContract>`?

Reply:
I don't think it's as simple as this. GoogleStorage is a `const`, so we can define a contract and use it in mocks: `export type GoogleStorageContract = ReturnType<typeof GoogleStorage>`

However, `analysis` is a function that returns an object, so to use this `Partial` syntax we would need to define a type that enumerates all properties of the returned `analysis` object, then type the function itself. This seems like working for typescript rather than having typescript work for us... Nonetheless, I was curious as a syntax learning exercise to do the work needed, so I included it and removed the `expect-errors`.

In conclusion, I added types to a good bit more of `GoogleStorage` file, specifically the `analysis` submodule, and defined a contract for that function.